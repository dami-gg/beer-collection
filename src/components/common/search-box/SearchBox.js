// @flow
import React, { Component } from "react";

import "./search-box.scss";

type State = {
  value: string
};

class SearchBox extends Component {
  state: State;
  onChange: Function;
  clearInput: Function;

  props: {
    className: string,
    changeHandler: Function
  };

  constructor(props: Object) {
    super(props);

    this.state = {
      value: ""
    };

    this.onChange = this.onChange.bind(this);
    this.clearInput = this.clearInput.bind(this);
  }

  onChange(event: Object): void {
    const value = event.target.value;

    this.setState({ value });
    this.props.changeHandler(value);
  }

  inputIsDirty(): boolean {
    return this.state.value !== "";
  }

  clearInput(): void {
    this.setState({ value: "" });
    this.props.changeHandler("");
  }

  render() {
    return (
      <div className={`search-box ${this.props.className}`}>
        <input
          className="search-box__input"
          type="text"
          value={this.state.value}
          placeholder="Search by name"
          onChange={this.onChange}
        />
        <div className="search-box__clear">
          {this.inputIsDirty() &&
            <i
              className="search-box__clear__button fa fa-times"
              onClick={this.clearInput}
            />}
        </div>
        <i className="search-box__icon fa fa-search" />
      </div>
    );
  }
}

export default SearchBox;
