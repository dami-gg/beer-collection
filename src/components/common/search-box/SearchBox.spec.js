import React from "react";
import SearchBox from "./SearchBox";

import { shallow } from "enzyme";
import renderer from "react-test-renderer";

describe("SearchBox", () => {
  let component, element, mock;

  beforeEach(mocks);

  it("should render without crashing", () => {
    shallow(<SearchBox />);
  });

  it("should match the snapshot", () => {
    component = renderer.create(<SearchBox />);
    let json = component.toJSON();

    expect(json).toMatchSnapshot();
  });

  describe("render", () => {
    beforeEach(setup);

    it("should render an input which on change calls the change handler", () => {
      element = component.find("input");
      element.simulate("change", mock.event);

      expect(mock.props.changeHandler).toHaveBeenCalled();
    });
  });

  function setup() {
    render();
  }

  function render() {
    component = shallow(<SearchBox changeHandler={mock.props.changeHandler} />);
  }

  function mocks() {
    mock = {
      props: {
        changeHandler: jest.fn()
      },
      event: {
        target: {
          value: ''
        }
      }
    };
  }
});
