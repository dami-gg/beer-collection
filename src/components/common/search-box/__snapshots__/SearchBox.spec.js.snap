exports[`SearchBox should match the snapshot 1`] = `
<div
  className="search-box">
  <input
    className="search-box__input"
    onChange={undefined}
    placeholder="Search by name"
    type="text" />
  <i
    className="search-box__icon fa fa-search" />
</div>
`;
